src/tachimawari/joint/utils/middleware.cpp

```diff
@@ -0,0 +1,218 @@
+{
+  if (control_rule == DEFAULT && control_type != FORCE) {
+    if (control_type == FOR_ACTION) {
+      action_counter++;
+    } else if (!is_action_controlling) {
+      if (control_type == FOR_HEAD) {
+        head_counter++;
+      } else if (control_type == FOR_WALKING) {
+        walking_counter++;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+std::vector<Joint> Middleware::filter_joints(
+  int control_type,
+  const std::vector<tachimawari_interfaces::msg::Joint> & joints_message)
+{
+  if (control_rule == DEFAULT || control_type == FORCE) {
+    std::vector<Joint> joints;
+    std::transform(
+      joints_message.begin(), joints_message.end(),
+      std::back_inserter(joints),
+      [](tachimawari_interfaces::msg::Joint joint) -> Joint {
+        return Joint(joint.id, joint.position);
+      });
+
+    if (control_type == FOR_ACTION) {
+      return joints;
+    } else if (!is_action_controlling || control_type == FORCE) {
+      return joints;
+    }
+  } else {
+    std::vector<Joint> joints;
+
+    if (control_type == FOR_ACTION) {
+      for (const auto & joint : joints_message) {
+        if (std::find(action_ids.begin(), action_ids.end(), joint.id) != action_ids.end()) {
+          joints.push_back(Joint(joint.id, joint.position));
+        }
+      }
+    } else if (control_type == FOR_HEAD) {
+      for (const auto & joint : joints_message) {
+        if (std::find(head_ids.begin(), head_ids.end(), joint.id) != head_ids.end()) {
+          joints.push_back(Joint(joint.id, joint.position));
+        }
+      }
+    } else if (control_type == FOR_WALKING) {
+      for (const auto & joint : joints_message) {
+        if (std::find(walking_ids.begin(), walking_ids.end(), joint.id) != walking_ids.end()) {
+          joints.push_back(Joint(joint.id, joint.position));
+        }
+      }
+    }
+
+    return joints;
+  }
+
+  return {};
+}
+
+void Middleware::update()
+{
+  if (action_counter == action_value) {
+    action_timer += time_unit;
+
+    if (action_timer > time_limit) {
+      is_action_controlling = false;
+      action_counter = 0;
+      action_value = 0;
+    }
+  } else {
+    action_timer = 0.0;
+    action_value = action_counter;
+    is_action_controlling = true;
+
+    if (control_rule == FOR_ACTION) {
+      reset_ids();
+    }
+  }
+
+  if (head_counter == head_value) {
+    head_timer += time_unit;
+
+    if (head_timer > time_limit) {
+      is_head_controlling = false;
+      head_counter = 0;
+      head_value = 0;
+    }
+  } else {
+    head_timer = 0.0;
+    head_value = head_counter;
+    is_head_controlling = true;
+  }
+
+  if (walking_counter == walking_value) {
+    walking_timer += time_unit;
+
+    if (walking_timer > time_limit) {
+      is_walking_controlling = false;
+      walking_counter = 0;
+      walking_value = 0;
+    }
+  } else {
+    walking_timer = 0.0;
+    walking_value = walking_counter;
+    is_walking_controlling = true;
+  }
```

Comment:
since action, head, and walking have the same behavior, let's create a class out of it